### PR TITLE
Feature testing

### LINKED ISSUE #120: [FEATURE] Update Help command to be able to take parameters and a nicer help message
**Is your feature request related to a problem? Please describe.**
Right now the help command only sends a message to the user and send the channel a message notifying that a help message was sent.

**Describe the solution you'd like**
We can improve this so that the command will be able to take a parameter to a command, and reply with the command name, description, and how to use it. We can also give every command a parameter where we can specify how to use each command.

**Describe alternatives you've considered**
We can at least make the help command instead of sending such a long list with description, just send a comma seperated list of all the commands into the channel the user is using it at. Then they can use the `help [command]` feature to find out more about each command

**Additional context**
This will make the help statemenet for the bot be a lot cleaner and nicer, that the user can use easily.

### PR DESCRIPTION
**Pull request by JoshQuaintance**
**Description:**
The help command now take parameters. If it doesn't recieve any, it will send in a list of all the commands. If the command receives a valid command, it will find it in the command array, and then send the user the usage and the description for the command

Closes #120 

- [x] Takes parameter
- [x] Shows list of all commands if param is not specified
- [x] Sends the usage and description for commands
- [x] Added a new value to each command: `usage` that will be used to display usage.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -3,6 +3,7 @@
   "version": "0.0.0",
   "private": true,
   "scripts": {
+    "prestart": "npm run build",
```
Comment (bradtaniguchi) [inline package.json]: How come this was added?

File: package.json [hunk 2]
```diff
@@ -22,7 +23,8 @@
     "mongoose": "^5.9.20",
     "node-fetch": "^2.6.0",
     "prettier": "^2.0.5",
-    "strip-ansi": "^6.0.0"
+    "strip-ansi": "^6.0.0",
+    "winston": "^3.3.3"
```
Comment (bradtaniguchi) [inline package.json]: I think you were also testing out changes for #119 and this got mixed in? 
You can revert this change by uninstalling this package, and or reverting the package.json and package-lock.json files.

File: src/commands/close.ts [hunk 3]
```diff
@@ -7,6 +7,7 @@ export const closeCommand: CommandDef = {
     'Closes the channel. This command requires admin privileges, ' +
     'and will only work on the automatically created "suspended" channels.' +
     ' Include the user if you want to remove the suspended role.',
+  usage: 'close',
```
Comment (bradtaniguchi) [inline src/commands/close.ts]: I think this command actually takes an argument for a "reason"? Might need to check with @nhcarrigan since he wrote this and I don't remember off the top of my head. (I also am being lazy right now and don't want to go thru the code, sorry :'( )

File: src/commands/eightball.ts [hunk 4]
```diff
@@ -5,6 +5,7 @@ export const eightBall: CommandDef = {
   prefix: 'eightball',
   description:
     'Think of your question and get an answer from a magic pool ball!',
+  usage: `\`eightball [message]\``,
```
Comment (bradtaniguchi) [inline src/commands/eightball.ts]: This is incorrect, this command doesn't take a message. This might of been done first and just forgotten about.
Please change to:
```suggestion
  usage: 'eightball',
```

File: src/commands/help.ts [hunk 5]
```diff
@@ -1,37 +1,63 @@
...
+      if (!commandFound)
+        return message.channel.send(
+          `Cannot find the command you specified -> \`${param}\``
+        );
+      const helpEmbed = new MessageEmbed()
+        .setColor('#0099FF')
+        .setTitle(commandFound.usage)
+        .addField('Description', `\`${commandFound.description}\``)
+        .addField('\u200b', 'Required: `<>` | Optional: `[]`');
+      message.channel.send(helpEmbed);
```
Comment (bradtaniguchi) [inline src/commands/help.ts]: The bot should send messages to the user in a DM rather than in chat, to prevent spam.

File: src/commands/help.ts [hunk 6]
```diff
@@ -1,37 +1,63 @@
...
+          'Bot Information',
+          'Hello! I am a test bot created by bradtaniguchi and members of ' +
+            'FreeCodeCamp.org to experiment with the process of building a ' +
+            'Discord Bot. You can view my [source code here](https://github.com/bradtaniguchi/discord-bot-test)'
+        )
+        // .addField(
+        //   'Commands',
+        //   `Hey there ${message.author}, here is the full list of all my commands.\n` +
+        //     `For more information about each command, use \`${config.PREFIX} help <command>\``
+        // )
```
Comment (bradtaniguchi) [inline src/commands/help.ts]: remove comments

File: src/commands/help.ts [hunk 7]
```diff
@@ -1,37 +1,63 @@
...
-        name: commandDef.prefix,
-        value: commandDef.description.substring(0, 1023),
-        inline: false
-      });
-    });
-    await message.author.send({ embed: helpEmbed });
-    message.channel.send('Help message sent!');
+  command: async (message, { config }) => {
+    const param = message.content.split(' ')[2];
+    if (param == undefined) {
```
Comment (bradtaniguchi) [inline src/commands/help.ts]: - Change `param` to `command`
- Change `if (param == undefined)` to `if(!param)` and **move the logic from the else to this if**.
- Move the logic in this command function to 2 separate functions, rather than if/else. Pass along any extra params in a single argument (similar to how the 2nd argument of the `command-def` works)

File: package.json (hunk 1 above)
Comment (JoshQuaintance) [reply package.json]: Owh, I added that myself, I just forgot to remove it again. It's just for my convinience so it will always  build before starting the bot. But I'll remove it.

File: src/commands/close.ts (hunk 3 above)
Comment (JoshQuaintance) [reply src/commands/close.ts]: Alright, I'll make sure to update it.

Comment (naomi-lgbt) [reply src/commands/close.ts]: Correct - the `close` command takes an optional `user` parameter (in the form of an @ mention). If the parameter is provided, that user is unsuspended.

File: src/commands/user.ts [hunk 8]
```diff
@@ -4,7 +4,8 @@ import { MessageEmbed } from 'discord.js';
 
 export const userCommand: CommandDef = {
   prefix: 'user',
-  description: "Get data on a current user. Use the syntax 'user [username]'.",
+  description: 'Get data on a given user, or the command caller themselves.',
```
Comment (naomi-lgbt) [inline src/commands/user.ts]: This is incorrect - the `user` command will not work if a user is not provided. 
This was done intentionally, as the `user` command is currently intended to tell a moderator if a user has previously been suspended. There is no need for a moderator to check this information on their own username :)

File: src/commands/suspend.ts [hunk 9]
```diff
@@ -5,8 +5,8 @@ import { userSuspendModel, UserSuspend } from '../APIs/mongo-suspend';
 export const suspendCommand: CommandDef = {
   prefix: 'suspend',
   description:
-    'Suspends a user for the given reason. This command is only available to admins. ' +
-    'Use the format "suspend <usertag> <reason>"',
+    'Suspends a user for the given reason. This command is only available to admins. ',
+  usage: 'suspend <usertag> <reason>',
```
Comment (naomi-lgbt) [inline src/commands/suspend.ts]: The `reason` parameter is optional - the bot will use "No reason provided" if this parameter is void.

File: src/commands/help.ts (hunk 5 above)
Comment (naomi-lgbt) [reply src/commands/help.ts]: The original command sent the embed directly to the author, and a "Help message sent" confirmation to the channel (to prevent the bot appearing to be unresponsive to other users).

**Review by naomi-lgbt — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (bradtaniguchi) [reply package.json]: Just to check, during development you can run `npm run dev`, which will run and live-reload the app. :)

File: src/commands/user.ts (hunk 8 above)
Comment (JoshQuaintance) [reply src/commands/user.ts]: Owh, so this is not like the user command that gives information about the user.

Comment (JoshQuaintance) [reply src/commands/user.ts]: Fixing it

Comment (naomi-lgbt) [reply src/commands/user.ts]: Yeh - it may be refactored later, but currently it only gets the prior suspension data.

File: src/commands/suspend.ts (hunk 9 above)
Comment (JoshQuaintance) [reply src/commands/suspend.ts]: Fixed

File: src/commands/coc.ts [hunk 10]
```diff
@@ -4,9 +4,10 @@ import { CommandDef } from './command-def';
 export const coc: CommandDef = {
   prefix: 'coc',
   description: 'Provides the Code of Conduct.',
+  usage: 'coc',
   command: (message): void => {
     const codeEmbed = new MessageEmbed()
-      .setTitle('FreeCodeCamp Code of Conduct')
+      .setTitle('freeCodeCamp Code of Conduct')
```
Comment (naomi-lgbt) [inline src/commands/coc.ts:10]: Good catch. XD

File: src/commands/help.ts [hunk 11]
```diff
@@ -1,37 +1,88 @@
...
-    });
-    await message.author.send({ embed: helpEmbed });
-    message.channel.send('Help message sent!');
+  command: async (message, { config }) => {
+    const Command = message.content.split(' ')[2];
+    if (!Command) {
+      sendFullList({ message, config });
+    } else {
+      sendSpecificCommand({ message, Command });
+    }
```
Comment (naomi-lgbt) [inline src/commands/help.ts]: Can leverage early returns to remove the `else`. @bradtaniguchi hates `else` XD

File: src/commands/quote.ts [hunk 12]
```diff
@@ -4,7 +4,9 @@ import { MotivationalDef } from '../APIs/quote-def';
 
 export const quote: CommandDef = {
   prefix: 'quote',
-  description: 'Returns a quote from the FCC motivational quotes file',
+  description:
+    "Returns a quote fetched from freeCodeCamp's motivational quotes API",
```
Comment (naomi-lgbt) [inline src/commands/quote.ts]: To clarify, it's fetched from the FCC `motivational-quotes.json` file, but the fetch uses the **GitHub** API. :)

File: src/commands/user.ts [hunk 13]
```diff
@@ -4,21 +4,18 @@ import { MessageEmbed } from 'discord.js';
 
 export const userCommand: CommandDef = {
   prefix: 'user',
-  description: "Get data on a current user. Use the syntax 'user [username]'.",
+  description: 'Get Data on the user mentioned.',
```
Comment (naomi-lgbt) [inline src/commands/user.ts]: Why is `Data` capitalised here? 🤔

File: src/commands/help.ts [hunk 14]
```diff
@@ -1,37 +1,88 @@
...
+    );
+  const helpEmbed = new MessageEmbed()
+    .setColor('#0099FF')
+    .setTitle(`\`${commandFound.usage}\``)
+    .addField('Description', commandFound.description)
+    .addField('\u200b', 'Required: `<>` | Optional: `[]`');
+
+  message.author.send(helpEmbed);
+  message.channel.send(
+    'Your help message with the command you requested is sent through DM'
```
Comment (naomi-lgbt) [inline src/commands/help.ts]:
```suggestion
    `DM sent with information on the ${commandFound.prefix} command`
```

Slight wording change.

File: src/commands/help.ts [hunk 15]
```diff
@@ -1,37 +1,88 @@
...
+        stripIndent`
+        \u200b
+        \u200b
+        For more information about each command, use \`${
+          (config as Config).PREFIX
+        } help <command>\``
+    )
+    .setFooter('I am not affiliated with FreeCodeCamp in any way.');
+  message.author.send(helpEmbed);
+  message.channel.send('Your help message is sent through DM');
```
Comment (naomi-lgbt) [inline src/commands/help.ts]:
```suggestion
  message.channel.send('DM sent with the requested information');
```

Slight wording change. Keeps it a little cleaner than using the second-person perspective.

**Review by naomi-lgbt — CHANGES_REQUESTED** [review]
Some more thoughts I had.

File: src/commands/user.ts (hunk 13 above)
Comment (JoshQuaintance) [reply src/commands/user.ts]: I didn't even know.

File: src/commands/eightball.ts [hunk 16]
```diff
@@ -5,6 +5,7 @@ export const eightBall: CommandDef = {
   prefix: 'eightball',
   description:
     'Think of your question and get an answer from a magic pool ball!',
+  usage: `eightball [message]`,
```
Comment (naomi-lgbt) [inline src/commands/eightball.ts]: Last one - remove this parameter and then LGTM

**Review by naomi-lgbt — CHANGES_REQUESTED** [review]

**Review by naomi-lgbt — APPROVED** [review]
LGTM!
